src/ray/gcs/gcs_server/test/gcs_server_rpc_test.cc

```diff
@@ -1,93 +1,218 @@
-      send_reply_callback(Status::OK(), nullptr, nullptr);
-    }
+  static void TearDownTestCase() {
+    std::string stop_redis_command = redis_client_executable + " -p 6379 shutdown";
+    RAY_LOG(INFO) << "Stop redis command is:" << stop_redis_command;
+    RAY_CHECK(system(stop_redis_command.c_str()) == 0);
+  }
+
+  void SetUp() override {
+    gcs::GcsServerConfig config;
+    config.grpc_server_port = 0;
+    config.grpc_server_name = "MockedGcsServer";
+    config.grpc_server_thread_num = 1;
+    config.redis_address = "127.0.0.1";
+    config.is_test = true;
+    gcs_server_.reset(new gcs::GcsServer(config));
+
+    thread_io_service_.reset(new std::thread([this] {
+      std::unique_ptr<boost::asio::io_service::work> work(
+          new boost::asio::io_service::work(io_service_));
+      io_service_.run();
+    }));
+
+    thread_gcs_server_.reset(new std::thread([this] { gcs_server_->Start(); }));
 
-    void HandleMarkJobFinished(const rpc::MarkJobFinishedRequest &request,
-                               rpc::MarkJobFinishedReply *reply,
-                               rpc::SendReplyCallback send_reply_callback) override {
-      send_reply_callback(Status::OK(), nullptr, nullptr);
+    // Wait until server starts listening.
+    while (gcs_server_->GetPort() == 0) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(10));
     }
-  };
-
-  gcs::GcsServerConfig config;
-  config.grpc_server_port = 0;
-  config.grpc_server_name = "MockedGcsServer";
-  config.grpc_server_thread_num = 1;
-  MockedGcsServer<MockedJobInfoHandler> server(config);
-  std::thread([&server] { server.Start(); }).detach();
-
-  // Wait until server starts listening.
-  while (server.GetPort() == 0) {
-    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+
+    // create gcs rpc client
+    client_call_manager_.reset(new rpc::ClientCallManager(io_service_));
+    client_.reset(
+        new rpc::GcsRpcClient("0.0.0.0", gcs_server_->GetPort(), *client_call_manager_));
   }
 
-  boost::asio::io_context client_io_context;
-  boost::asio::io_context::work worker(client_io_context);
-  rpc::ClientCallManager client_call_manager(client_io_context);
-  rpc::GcsRpcClient client("0.0.0.0", server.GetPort(), client_call_manager);
-  std::thread([&client_io_context] { client_io_context.run(); }).detach();
+  void TearDown() override {
+    gcs_server_->Stop();
+    io_service_.stop();
+    thread_io_service_->join();
+    thread_gcs_server_->join();
+  }
 
+  void TestAddJob(const rpc::AddJobRequest &request) {
+    std::promise<rpc::AddJobReply> reply_promise;
+    auto reply_future = reply_promise.get_future();
+    client_->AddJob(
+        request, [&reply_promise](const Status &status, const rpc::AddJobReply &reply) {
+          if (status.ok()) {
+            reply_promise.set_value(reply);
+          }
+        });
+    auto future_status = reply_future.wait_for(std::chrono::milliseconds(200));
```

Comment:
define `200` as a constant. Also 200ms might not be enough for slow machines. I think it's fine to make this larger, e.g., 2s.

Reply:
fixed